addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,607 @@
+	dynClient, err := dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		r.Log.Error(err, "Error creating dynamic client")
+		return err
+	}
+	r.dynamicClient = dynClient
+	r.providerWatches = make(map[string]client.Object)
+
+	r.providerGVR = make(map[schema.GroupKind]*schema.GroupVersionResource)
+	clientset := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+	r.cachedDiscoveryClient = cacheddiscovery.NewMemCacheClient(clientset.Discovery())
+
+	go r.periodicGVRCachesClean()
+
+	return nil
+}
+
+// Reconcile performs the reconciliation action for the controller.
+func (r *tanzuClusterBootstrapReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues(constants.ClusterNamespaceLogKey, req.Namespace, constants.ClusterNameLogKey, req.Name)
+	// get cluster object
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	tkrName := util.GetTKRNameForCluster(r.context, r.Client, cluster)
+	if tkrName == "" {
+		log.Info("cluster does not have an associated TKR")
+		return ctrl.Result{}, nil
+	}
+	// process only v1alpha3
+	tkr := &runtanzuv1alpha3.TanzuKubernetesRelease{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: constants.TKGBomNamespace, Name: tkrName}, tkr); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	log.Info("Reconciling cluster")
+
+	// if deletion timestamp is set, handle cluster deletion
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		// TODO delete
+	} else {
+		// reconcile addons in cluster
+		result, err := r.reconcileNormal(cluster, log)
+		if err != nil {
+			log.Error(err, "failed to reconcile cluster")
+			return ctrl.Result{}, err
+		}
+
+		return result, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileNormal reconciles the TanzuBootstrapCluster object
+func (r *tanzuClusterBootstrapReconciler) reconcileNormal(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
+	// get or clone or patch from template
+	tanzuClusterBootstrap, err := r.createOrPatchTanzuClusterBootstrapFromTemplate(cluster, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if tanzuClusterBootstrap == nil {
+		return ctrl.Result{}, nil
+	}
+
+	remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+	if err != nil {
+		log.Error(err, "Error getting remote cluster client")
+		return ctrl.Result{}, err
+	}
+	// TODO handle kapp as a remote packageinstall from mgmt cluster
+	// TODO handle cni, cpi, csi
+	// handle additionalPackages
+	for _, additionalPkg := range tanzuClusterBootstrap.Spec.AdditionalPackages {
+		// TODO packageinstall (rbac, serviceaccount, package, packageinstall)
```

Comment:
question: what happens if user has already created legacy addon `Secret` and so there is already a `PackageInstall` for that package? (i know this is outside the bounds of this PR, it was just a question I had when trying to figure out the pinniped flow)

Reply:
Short answer: We'll use annotations on secrets to figure out conflicts. 

For upgrade in general there are various aspects to upgrade. I think the one to address will be upgrade from TKGm 1.6 to uTKG for AWS and Azure. For vsphere it will be a migration. @danniel1205 has started collecting some thoughts on a doc so it would be good to chat in a sync meeting with @vuil and @imikushin on their thoughts as well.